src/types.c

```diff
@@ -21,13 +21,16 @@ enum vctrs_type vec_typeof(SEXP x) {
   case STRSXP: return vctrs_type_character;
   case RAWSXP: return vctrs_type_raw;
   case VECSXP:
+    if (!OBJECT(x)) {
+      return vctrs_type_list;
+    }
```

Comment:
I feel like this should be an `else if`?

Reply:
I thought the early return makes clearer code with less nestedness:

```c
    if (OBJECT(x)) {
      if (is_data_frame(x)) {
        return vctrs_type_dataframe;
      } else {
        return vctrs_type_s3;
      }
    } else {
      return vctrs_type_list;
    }
```

versus

```c
    if (!OBJECT(x)) {
      return vctrs_type_list;
    }
    if (is_data_frame(x)) {
      return vctrs_type_dataframe;
    } else {
      return vctrs_type_s3;
    }
```

The S3 branch will get more complicated in the future so it seems better to unnest it.

Reply:
I just meant like this:

```c
if (!OBJECT(x)) {
  return vctrs_type_list;
} else if (is_data_frame(x)) {
  return vctrs_type_dataframe;
} else {
  return vctrs_type_s3;
}
```

Reply:
But it's not a big deal.

Reply:
Ah yes, done.